src/thor/triplegbuilder.cc

```diff
@@ -91,93 +91,190 @@ void AssignAdmins(const AttributesController& controller,
+                        const std::vector<baldr::GraphReader::Incident>& incidents) {
   // TODO: if this is a transit edge then the costing will throw
 
-  if (trip_path.has_shape_attributes()) {
-    // A list of percent along the edge and corresponding speed (meters per second)
-    std::vector<std::tuple<double, double>> speeds;
-    double speed = (edge->length() * (tgt_pct - src_pct)) / edge_seconds;
-    if (cut_for_traffic) {
-      // TODO: we'd like to use the speed from traffic here but because there are synchronization
-      // problems with those records changing between when we used them to make the path and when we
-      // try to grab them again here, we instead rely on the total time from PathInfo and just do the
-      // cutting for now
-      const auto& traffic_speed = tile->trafficspeed(edge);
-      if (traffic_speed.breakpoint1 > 0) {
-        speeds.emplace_back(traffic_speed.breakpoint1 / 255.0, speed);
-        if (traffic_speed.breakpoint2 > 0) {
-          speeds.emplace_back(traffic_speed.breakpoint2 / 255.0, speed);
-          if (traffic_speed.speed3 != UNKNOWN_TRAFFIC_SPEED_RAW) {
-            speeds.emplace_back(1, speed);
-          }
+  // bail if nothing to do
+  if (!cut_for_traffic && incidents.empty() &&
+      !controller.category_attribute_enabled(kShapeAttributesCategory))
+    return;
+
+  // initialize shape_attributes once
+  if (!leg.has_shape_attributes() &&
+      controller.category_attribute_enabled(kShapeAttributesCategory)) {
+    leg.mutable_shape_attributes();
+  }
+
+  // convenient for bundling info about the spots where we cut the shape
+  struct cut_t {
+    double percent_along;
+    double speed; // meters per second
+    uint8_t congestion;
+    std::vector<const baldr::GraphReader::Incident*> incidents;
+  };
+
+  // A list of percent along the edge, corresponding speed (meters per second), incident id
+  std::vector<cut_t> cuts;
+  double speed = (edge->length() * (tgt_pct - src_pct)) / edge_seconds;
+  if (cut_for_traffic) {
+    // TODO: we'd like to use the speed from traffic here but because there are synchronization
+    // problems with those records changing between when we used them to make the path and when we
+    // try to grab them again here, we instead rely on the total time from PathInfo and just do the
+    // cutting for now
+    const auto& traffic_speed = tile->trafficspeed(edge);
+    if (traffic_speed.breakpoint1 > 0) {
+      cuts.emplace_back(cut_t{traffic_speed.breakpoint1 / 255.0, speed,
+                              static_cast<std::uint8_t>(traffic_speed.congestion1)});
+      if (traffic_speed.breakpoint2 > 0) {
+        cuts.emplace_back(cut_t{traffic_speed.breakpoint2 / 255.0, speed,
+                                static_cast<std::uint8_t>(traffic_speed.congestion2)});
+        if (traffic_speed.speed3 != UNKNOWN_TRAFFIC_SPEED_RAW) {
+          cuts.emplace_back(cut_t{1, speed, static_cast<std::uint8_t>(traffic_speed.congestion3)});
```

Comment:
this bit here hasnt changed, other than `tuple->cut_t`, it just got pulled out of an `if` so it shows as a diff

Reply:
Switching to "ignore whitespace changes" in diff makes this much easier to follow